contracts/lib.rs

```diff
@@ -787,90 +795,145 @@ pub mod prosopo {
             };
-            let dapp = Dapp {
-                status,
-                balance: transferred,
+        }
+
+        /// Configure a dapp (existing or new)
+        fn dapp_configure(
+            &mut self,
+            contract: AccountId,
+            payee: DappPayee,
+            owner: AccountId,
+        ) -> Result<Dapp, Error> {
+            self.check_is_contract(contract)?;
+
+            let dapp_lookup = self.dapps.get(contract);
+            let new = dapp_lookup.is_none();
+            let mut dapp = dapp_lookup.unwrap_or(Dapp {
                 owner,
-                min_difficulty: 1,
+                balance: 0,
+                status: GovernanceStatus::Suspended,
                 payee,
-            };
-            // keying on contract allows owners to own many contracts
+                min_difficulty: 1,
+            });
+
+            // check current contract for ownership
+            if !new {
+                self.check_dapp_owner_is_caller(contract)?;
+            }
+
+            // update the dapp payee
+            dapp.payee = payee;
+
+            // update the owner
+            dapp.owner = owner;
+
+            self.dapp_configure_funding(&mut dapp);
+
+            // if the dapp is new then add it to the list of dapps
+            if new {
+                lazy_push!(self.dapp_accounts, contract);
+            }
+
+            // update the dapp in the mapping
             self.dapps.insert(contract, &dapp);
-            let mut dapp_accounts = self.dapp_accounts.get_or_default();
-            dapp_accounts.push(contract);
-            self.dapp_accounts.set(&dapp_accounts);
+
+            Ok(dapp)
+        }
+
+        /// Register a dapp
+        #[ink(message)]
+        pub fn dapp_register(
+            &mut self,
+            contract: AccountId,
+            payee: DappPayee,
+        ) -> Result<(), Error> {
+            // expect dapp to be new
+            self.check_dapp_does_not_exist(contract)?;
+
+            // configure the new dapp
+            let dapp = self.dapp_configure(
+                contract,
+                payee,
+                self.env().caller(), // the caller is made the owner of the contract
+            )?;
+
             // emit event
             self.env().emit_event(DappRegister {
                 contract,
-                owner,
-                value: transferred,
+                owner: dapp.owner,
+                value: dapp.balance,
                 payee,
-                status,
+                status: dapp.status,
             });
 
             Ok(())
         }
 
         /// Update a dapp with new funds, setting status as appropriate
-        fn dapp_update(
+        pub fn dapp_update(
             &mut self,
-            owner: AccountId,
-            transferred: u128,
             contract: AccountId,
-            caller: AccountId,
             payee: DappPayee,
+            owner: AccountId,
```

Comment:
Surely owner is not required since we are forcing the caller to be the owner at registration. `self.env.caller()` could be used instead, same as registration.

Reply:
very true, I'd missed that. I'll fix it :)

Reply:
owner is the new owner here, so if I were transferring dapp A from person B to person C, owner would be person C's address

Reply:
I suppose this works with the pending requirement of the owner being the dapp deployer, as the original owner has to do the update.